// @medusajs/medusa/dist/models/payment-collection.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentCollection = exports.PaymentCollectionType = exports.PaymentCollectionStatus = void 0;
var typeorm_1 = require("typeorm");
var soft_deletable_entity_1 = require("../interfaces/models/soft-deletable-entity");
var db_aware_column_1 = require("../utils/db-aware-column");
var utils_1 = require("../utils");
var _1 = require(".");
var order_editing_1 = __importDefault(require("../loaders/feature-flags/order-editing"));
var feature_flag_decorators_1 = require("../utils/feature-flag-decorators");
var PaymentCollectionStatus;
(function (PaymentCollectionStatus) {
    PaymentCollectionStatus["NOT_PAID"] = "not_paid";
    PaymentCollectionStatus["AWAITING"] = "awaiting";
    PaymentCollectionStatus["AUTHORIZED"] = "authorized";
    PaymentCollectionStatus["PARTIALLY_AUTHORIZED"] = "partially_authorized";
    PaymentCollectionStatus["CAPTURED"] = "captured";
    PaymentCollectionStatus["PARTIALLY_CAPTURED"] = "partially_captured";
    PaymentCollectionStatus["REFUNDED"] = "refunded";
    PaymentCollectionStatus["PARTIALLY_REFUNDED"] = "partially_refunded";
    PaymentCollectionStatus["CANCELED"] = "canceled";
    PaymentCollectionStatus["REQUIRES_ACTION"] = "requires_action";
})(PaymentCollectionStatus = exports.PaymentCollectionStatus || (exports.PaymentCollectionStatus = {}));
var PaymentCollectionType;
(function (PaymentCollectionType) {
    PaymentCollectionType["ORDER_EDIT"] = "order_edit";
})(PaymentCollectionType = exports.PaymentCollectionType || (exports.PaymentCollectionType = {}));
var PaymentCollection = /** @class */ (function (_super) {
    __extends(PaymentCollection, _super);
    function PaymentCollection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PaymentCollection.prototype.beforeInsert = function () {
        this.id = (0, utils_1.generateEntityId)(this.id, "paycol");
    };
    __decorate([
        (0, db_aware_column_1.DbAwareColumn)({ type: "enum", enum: PaymentCollectionType }),
        __metadata("design:type", String)
    ], PaymentCollection.prototype, "type", void 0);
    __decorate([
        (0, db_aware_column_1.DbAwareColumn)({ type: "enum", enum: PaymentCollectionStatus }),
        __metadata("design:type", String)
    ], PaymentCollection.prototype, "status", void 0);
    __decorate([
        (0, typeorm_1.Column)({ nullable: true }),
        __metadata("design:type", String)
    ], PaymentCollection.prototype, "description", void 0);
    __decorate([
        (0, typeorm_1.Column)({ type: "int" }),
        __metadata("design:type", Number)
    ], PaymentCollection.prototype, "amount", void 0);
    __decorate([
        (0, typeorm_1.Column)({ type: "int", nullable: true }),
        __metadata("design:type", Number)
    ], PaymentCollection.prototype, "authorized_amount", void 0);
    __decorate([
        (0, typeorm_1.Column)({ type: "int", nullable: true }),
        __metadata("design:type", Number)
    ], PaymentCollection.prototype, "refunded_amount", void 0);
    __decorate([
        (0, typeorm_1.Index)(),
        (0, typeorm_1.Column)(),
        __metadata("design:type", String)
    ], PaymentCollection.prototype, "region_id", void 0);
    __decorate([
        (0, typeorm_1.ManyToOne)(function () { return _1.Region; }),
        (0, typeorm_1.JoinColumn)({ name: "region_id" }),
        __metadata("design:type", _1.Region)
    ], PaymentCollection.prototype, "region", void 0);
    __decorate([
        (0, typeorm_1.Index)(),
        (0, typeorm_1.Column)(),
        __metadata("design:type", String)
    ], PaymentCollection.prototype, "currency_code", void 0);
    __decorate([
        (0, typeorm_1.ManyToOne)(function () { return _1.Currency; }),
        (0, typeorm_1.JoinColumn)({ name: "currency_code", referencedColumnName: "code" }),
        __metadata("design:type", _1.Currency)
    ], PaymentCollection.prototype, "currency", void 0);
    __decorate([
        (0, typeorm_1.ManyToMany)(function () { return _1.PaymentSession; }),
        (0, typeorm_1.JoinTable)({
            name: "payment_collection_sessions",
            joinColumn: {
                name: "payment_collection_id",
                referencedColumnName: "id",
            },
            inverseJoinColumn: {
                name: "payment_session_id",
                referencedColumnName: "id",
            },
        }),
        __metadata("design:type", Array)
    ], PaymentCollection.prototype, "payment_sessions", void 0);
    __decorate([
        (0, typeorm_1.ManyToMany)(function () { return _1.Payment; }),
        (0, typeorm_1.JoinTable)({
            name: "payment_collection_payments",
            joinColumn: {
                name: "payment_collection_id",
                referencedColumnName: "id",
            },
            inverseJoinColumn: {
                name: "payment_id",
                referencedColumnName: "id",
            },
        }),
        __metadata("design:type", Array)
    ], PaymentCollection.prototype, "payments", void 0);
    __decorate([
        (0, db_aware_column_1.DbAwareColumn)({ type: "jsonb" }),
        __metadata("design:type", Object)
    ], PaymentCollection.prototype, "metadata", void 0);
    __decorate([
        (0, typeorm_1.Column)(),
        __metadata("design:type", String)
    ], PaymentCollection.prototype, "created_by", void 0);
    __decorate([
        (0, typeorm_1.BeforeInsert)(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], PaymentCollection.prototype, "beforeInsert", null);
    PaymentCollection = __decorate([
        (0, feature_flag_decorators_1.FeatureFlagEntity)(order_editing_1.default.key)
    ], PaymentCollection);
    return PaymentCollection;
}(soft_deletable_entity_1.SoftDeletableEntity));
exports.PaymentCollection = PaymentCollection;
